File: src/runtime/DrawingCanvasGestureState.js
.pragma library

// Legacy parity reference for the C++ drawing_canvas_core runtime.
// App QML should call drawingCanvasRuntime instead of importing this file.

function finiteNumber(value, fallback) {
    var number = Number(value)
    return Number.isFinite(number) ? number : fallback
}

function asArray(value) {
    if (!value) {
        return []
    }
    if (Array.isArray(value)) {
        return value
    }
    if (typeof value.length === "number") {
        var result = []
        for (var index = 0; index < value.length; ++index) {
            result.push(value[index])
        }
        return result
    }
    return []
}

function normalizePoint(point) {
    return {
        x: finiteNumber(point && point.x, 0),
        y: finiteNumber(point && point.y, 0)
    }
}

function normalizeScreenPoint(point) {
    return {
        x: finiteNumber(point && point.x, 0),
        y: finiteNumber(point && point.y, 0)
    }
}

function normalizeModifiers(modifiers) {
    return {
        shift: modifiers && modifiers.shift === true,
        alt: modifiers && modifiers.alt === true,
        control: modifiers && modifiers.control === true,
        meta: modifiers && modifiers.meta === true
    }
}

function noneIntent() {
    return {
        kind: "none"
    }
}

function initialGestureState() {
    return {
        mode: "idle",
        started: false,
        objectId: "",
        handleId: "",
        selectedIds: [],
        startPoint: { x: 0, y: 0 },
        lastPoint: { x: 0, y: 0 },
        startScreenPoint: { x: 0, y: 0 },
        lastScreenPoint: { x: 0, y: 0 },
        moved: false,
        modifiers: normalizeModifiers({}),
        targetKind: "none",
        targetObjectId: "",
        targetHandleId: "",
        rejected: false
    }
}

function cloneState(state) {
    var source = state || initialGestureState()
    return {
        mode: String(source.mode || "idle"),
        started: source.started === true,
        objectId: String(source.objectId || ""),
        handleId: String(source.handleId || ""),
        selectedIds: asArray(source.selectedIds).map(function(id) { return String(id || "") }),
        startPoint: normalizePoint(source.startPoint),
        lastPoint: normalizePoint(source.lastPoint),
        startScreenPoint: normalizeScreenPoint(source.startScreenPoint),
        lastScreenPoint: normalizeScreenPoint(source.lastScreenPoint),
        moved: source.moved === true,
        modifiers: normalizeModifiers(source.modifiers),
        targetKind: String(source.targetKind || "none"),
        targetObjectId: String(source.targetObjectId || ""),
        targetHandleId: String(source.targetHandleId || ""),
        rejected: source.rejected === true
    }
}

function activeMode(mode) {
    return mode === "drawing_pending_shape"
            || mode === "dragging_object"
            || mode === "dragging_handle"
            || mode === "marquee_select"
            || mode === "panning"
}

function transitionAllowed(fromMode, toMode) {
    var from = String(fromMode || "idle")
    var to = String(toMode || "idle")
    if (from === to) {
        return true
    }
    if (to === "idle") {
        return true
    }
    if (from === "idle") {
        return true
    }
    if (from === "hovering" && !activeMode(to)) {
        return true
    }
    if (from === "hovering" && activeMode(to)) {
        return true
    }
    return false
}

function rejectedState(state) {
    var next = cloneState(state)
    next.rejected = true
    return next
}

function beginGesture(state, mode, payload) {
    var current = cloneState(state)
    var nextMode = String(mode || "idle")
    if (!transitionAllowed(current.mode, nextMode)) {
        return rejectedState(current)
    }
    var point = normalizePoint(payload && payload.point)
    var screenPoint = normalizeScreenPoint(payload && payload.screenPoint)
    return {
        mode: nextMode,
        started: nextMode !== "idle" && nextMode !== "hovering",
        objectId: String(payload && payload.objectId || ""),
        handleId: String(payload && payload.handleId || ""),
        selectedIds: asArray(payload && payload.selectedIds).map(function(id) { return String(id || "") }),
        startPoint: point,
        lastPoint: point,
        startScreenPoint: screenPoint,
        lastScreenPoint: screenPoint,
        moved: false,
        modifiers: normalizeModifiers(payload && payload.modifiers),
        targetKind: String(payload && payload.targetKind || "none"),
        targetObjectId: String(payload && payload.targetObjectId || ""),
        targetHandleId: String(payload && payload.targetHandleId || ""),
        rejected: false
    }
}

function beginHover(state, point, target) {
    return beginGesture(state, "hovering", {
        point: point,
        targetKind: target && target.kind || "none",
        targetObjectId: target && target.objectId || "",
        targetHandleId: target && target.handleId || "",
        modifiers: target && target.modifiers || {}
    })
}

function beginObjectDrag(state, objectId, point, selectedIds, modifiers) {
    return beginGesture(state, "dragging_object", {
        objectId: objectId,
        point: point,
        selectedIds: selectedIds,
        modifiers: modifiers
    })
}

function beginHandleDrag(state, objectId, handleId, point, modifiers) {
    return beginGesture(state, "dragging_handle", {
        objectId: objectId,
        handleId: handleId,
        point: point,
        modifiers: modifiers
    })
}

function beginMarquee(state, point, modifiers) {
    return beginGesture(state, "marquee_select", {
        point: point,
        modifiers: modifiers
    })
}

function beginPan(state, screenPoint, modifiers) {
    return beginGesture(state, "panning", {
        screenPoint: screenPoint,
        modifiers: modifiers
    })
}

function beginDrawingPendingShape(state, point, modifiers) {
    return beginGesture(state, "drawing_pending_shape", {
        point: point,
        modifiers: modifiers
    })
}

function updateGesture(state, payload) {
    var next = cloneState(state)
    next.rejected = false
    next.modifiers = normalizeModifiers(payload && payload.modifiers || next.modifiers)
    if (payload && payload.point) {
        next.lastPoint = normalizePoint(payload.point)
    }
    if (payload && payload.screenPoint) {
        next.lastScreenPoint = normalizeScreenPoint(payload.screenPoint)
    }
    if (payload && payload.targetKind) {
        next.targetKind = String(payload.targetKind || "none")
        next.targetObjectId = String(payload.targetObjectId || "")
        next.targetHandleId = String(payload.targetHandleId || "")
    }
    var dx = next.lastPoint.x - next.startPoint.x
    var dy = next.lastPoint.y - next.startPoint.y
    var sdx = next.lastScreenPoint.x - next.startScreenPoint.x
    var sdy = next.lastScreenPoint.y - next.startScreenPoint.y
    var tolerance = Math.max(0, finiteNumber(payload && payload.moveTolerance, 0.000001))
    var screenTolerance = Math.max(0, finiteNumber(payload && payload.screenMoveTolerancePx, 0))
    next.moved = next.moved
            || Math.abs(dx) > tolerance
            || Math.abs(dy) > tolerance
            || Math.abs(sdx) > screenTolerance
            || Math.abs(sdy) > screenTolerance
    return next
}

function finishIntent(state, payload) {
    var current = cloneState(state)
    if (payload && payload.incremental === true) {
        return noneIntent()
    }
    var finalPoint = payload && payload.point ? normalizePoint(payload.point) : current.lastPoint
    var finalScreenPoint = payload && payload.screenPoint ? normalizeScreenPoint(payload.screenPoint) : current.lastScreenPoint
    if (current.mode === "dragging_handle") {
        return {
            kind: "update_handle",
            objectId: current.objectId,
            handleId: current.handleId,
            point: finalPoint
        }
    }
    if (current.mode === "dragging_object") {
        var dx = finalPoint.x - current.startPoint.x
        var dy = finalPoint.y - current.startPoint.y
        if (current.selectedIds.length > 1) {
            return {
                kind: "move_objects",
                objectIds: current.selectedIds,
                dx: dx,
                dy: dy
            }
        }
        return {
            kind: "move_object",
            objectId: current.objectId,
            dx: dx,
            dy: dy
        }
    }
    if (current.mode === "marquee_select") {
        return {
            kind: "select_objects",
            objectIds: asArray(payload && payload.objectIds).map(function(id) { return String(id || "") }),
            startPoint: current.startPoint,
            endPoint: finalPoint
        }
    }
    if (current.mode === "panning") {
        return {
            kind: "pan",
            dxPx: finalScreenPoint.x - current.startScreenPoint.x,
            dyPx: finalScreenPoint.y - current.startScreenPoint.y
        }
    }
    if (current.mode === "drawing_pending_shape") {
        return {
            kind: "draw_click",
            point: finalPoint
        }
    }
    return noneIntent()
}

function finishGesture(state, payload) {
    return {
        state: initialGestureState(),
        intent: finishIntent(state, payload)
    }
}

function cancelGesture(state) {
    return {
        state: initialGestureState(),
        intent: noneIntent()
    }
}

function isDragging(state) {
    var mode = String(state && state.mode || "")
    return mode === "dragging_object" || mode === "dragging_handle"
}

function isHandleDrag(state) {
    return String(state && state.mode || "") === "dragging_handle"
}

function isObjectDrag(state) {
    return String(state && state.mode || "") === "dragging_object"
}

function isMarquee(state) {
    return String(state && state.mode || "") === "marquee_select"
}

function finishKind(state) {
    var current = cloneState(state)
    if (isHandleDrag(current) || isObjectDrag(current)) {
        return "incremental_drag"
    }
    if (isMarquee(current)) {
        return current.moved ? "marquee_select" : "marquee_click"
    }
    if (current.mode === "panning") {
        return "pan"
    }
    if (current.mode === "drawing_pending_shape") {
        return "draw_click"
    }
    return "none"
}

function finishAction(state) {
    var kind = finishKind(state)
    return {
        kind: kind,
        shouldFinishIncrementalDrag: kind === "incremental_drag",
        shouldSelectMarquee: kind === "marquee_select",
        shouldSuppressClick: kind === "incremental_drag" || kind === "marquee_select",
        shouldEndObjectMove: kind !== "marquee_select",
        shouldResetLifecycle: true
    }
}

function gestureLabel(state) {
    var mode = String(state && state.mode || "idle")
    if (mode === "dragging_handle") {
        return String(state && state.handleId || "") === "rect_rotate" ? "rotate" : "drag handle"
    }
    if (mode === "dragging_object") {
        return asArray(state && state.selectedIds).length > 1 ? "move selection" : "move object"
    }
    if (mode === "marquee_select") {
        return "marquee"
    }
    if (mode === "panning") {
        return "pan"
    }
    if (mode === "drawing_pending_shape") {
        return "draw"
    }
    return mode === "hovering" ? "hover" : "idle"
}

File: src/canvas/DrawingCanvasGestureState.cpp
#include "DrawingCanvasGestureState.h"

#include <cmath>

namespace drawing_canvas {
namespace {

QVariantList asStringListVariant(const QVariant &value) {
    QVariantList result;
    const QVariantList source = value.toList();
    for (const QVariant &entry : source) {
        result.push_back(entry.toString());
    }
    return result;
}

QVariantMap normalizePointMap(const QVariant &value) {
    return pointToVariant(pointFromVariant(value));
}

QVariantMap normalizeScreenPointMap(const QVariant &value) {
    const QVariantMap map = value.toMap();
    return {
        {QStringLiteral("x"), finiteNumber(map.value(QStringLiteral("x")), 0.0)},
        {QStringLiteral("y"), finiteNumber(map.value(QStringLiteral("y")), 0.0)}
    };
}

QVariantMap normalizeModifiers(const QVariant &value) {
    const QVariantMap map = value.toMap();
    return {
        {QStringLiteral("shift"), map.value(QStringLiteral("shift")).toBool()},
        {QStringLiteral("alt"), map.value(QStringLiteral("alt")).toBool()},
        {QStringLiteral("control"), map.value(QStringLiteral("control")).toBool()},
        {QStringLiteral("meta"), map.value(QStringLiteral("meta")).toBool()}
    };
}

QVariantMap noneIntent() {
    return {{QStringLiteral("kind"), QStringLiteral("none")}};
}

QVariantMap cloneState(const QVariantMap &state) {
    const QVariantMap source = state.isEmpty() ? initialGestureState() : state;
    return {
        {QStringLiteral("mode"), source.value(QStringLiteral("mode"), QStringLiteral("idle")).toString()},
        {QStringLiteral("started"), source.value(QStringLiteral("started")).toBool()},
        {QStringLiteral("objectId"), source.value(QStringLiteral("objectId")).toString()},
        {QStringLiteral("handleId"), source.value(QStringLiteral("handleId")).toString()},
        {QStringLiteral("selectedIds"), asStringListVariant(source.value(QStringLiteral("selectedIds")))},
        {QStringLiteral("startPoint"), normalizePointMap(source.value(QStringLiteral("startPoint")))},
        {QStringLiteral("lastPoint"), normalizePointMap(source.value(QStringLiteral("lastPoint")))},
        {QStringLiteral("startScreenPoint"), normalizeScreenPointMap(source.value(QStringLiteral("startScreenPoint")))},
        {QStringLiteral("lastScreenPoint"), normalizeScreenPointMap(source.value(QStringLiteral("lastScreenPoint")))},
        {QStringLiteral("moved"), source.value(QStringLiteral("moved")).toBool()},
        {QStringLiteral("modifiers"), normalizeModifiers(source.value(QStringLiteral("modifiers")))},
        {QStringLiteral("targetKind"), source.value(QStringLiteral("targetKind"), QStringLiteral("none")).toString()},
        {QStringLiteral("targetObjectId"), source.value(QStringLiteral("targetObjectId")).toString()},
        {QStringLiteral("targetHandleId"), source.value(QStringLiteral("targetHandleId")).toString()},
        {QStringLiteral("rejected"), source.value(QStringLiteral("rejected")).toBool()}
    };
}

bool activeMode(const QString &mode) {
    return mode == QStringLiteral("drawing_pending_shape")
        || mode == QStringLiteral("dragging_object")
        || mode == QStringLiteral("dragging_handle")
        || mode == QStringLiteral("marquee_select")
        || mode == QStringLiteral("panning");
}

bool transitionAllowed(const QString &fromMode, const QString &toMode) {
    const QString from = fromMode.isEmpty() ? QStringLiteral("idle") : fromMode;
    const QString to = toMode.isEmpty() ? QStringLiteral("idle") : toMode;
    if (from == to || to == QStringLiteral("idle") || from == QStringLiteral("idle")) {
        return true;
    }
    if (from == QStringLiteral("hovering")) {
        return true;
    }
    return false;
}

QVariantMap rejectedState(const QVariantMap &state) {
    QVariantMap next = cloneState(state);
    next.insert(QStringLiteral("rejected"), true);
    return next;
}

QVariantMap beginGesture(const QVariantMap &state, const QString &mode, const QVariantMap &payload) {
    const QVariantMap current = cloneState(state);
    const QString nextMode = mode.isEmpty() ? QStringLiteral("idle") : mode;
    if (!transitionAllowed(current.value(QStringLiteral("mode")).toString(), nextMode)) {
        return rejectedState(current);
    }
    const QVariantMap point = normalizePointMap(payload.value(QStringLiteral("point")));
    const QVariantMap screenPoint = normalizeScreenPointMap(payload.value(QStringLiteral("screenPoint")));
    return {
        {QStringLiteral("mode"), nextMode},
        {QStringLiteral("started"), nextMode != QStringLiteral("idle") && nextMode != QStringLiteral("hovering")},
        {QStringLiteral("objectId"), payload.value(QStringLiteral("objectId")).toString()},
        {QStringLiteral("handleId"), payload.value(QStringLiteral("handleId")).toString()},
        {QStringLiteral("selectedIds"), asStringListVariant(payload.value(QStringLiteral("selectedIds")))},
        {QStringLiteral("startPoint"), point},
        {QStringLiteral("lastPoint"), point},
        {QStringLiteral("startScreenPoint"), screenPoint},
        {QStringLiteral("lastScreenPoint"), screenPoint},
        {QStringLiteral("moved"), false},
        {QStringLiteral("modifiers"), normalizeModifiers(payload.value(QStringLiteral("modifiers")))},
        {QStringLiteral("targetKind"), payload.value(QStringLiteral("targetKind"), QStringLiteral("none")).toString()},
        {QStringLiteral("targetObjectId"), payload.value(QStringLiteral("targetObjectId")).toString()},
        {QStringLiteral("targetHandleId"), payload.value(QStringLiteral("targetHandleId")).toString()},
        {QStringLiteral("rejected"), false}
    };
}

QVariantMap finishIntent(const QVariantMap &state, const QVariantMap &payload) {
    const QVariantMap current = cloneState(state);
    if (payload.value(QStringLiteral("incremental")).toBool()) {
        return noneIntent();
    }
    const QVariantMap finalPoint = payload.contains(QStringLiteral("point"))
        ? normalizePointMap(payload.value(QStringLiteral("point")))
        : current.value(QStringLiteral("lastPoint")).toMap();
    const QVariantMap finalScreenPoint = payload.contains(QStringLiteral("screenPoint"))
        ? normalizeScreenPointMap(payload.value(QStringLiteral("screenPoint")))
        : current.value(QStringLiteral("lastScreenPoint")).toMap();
    const QString mode = current.value(QStringLiteral("mode")).toString();
    if (mode == QStringLiteral("dragging_handle")) {
        return {
            {QStringLiteral("kind"), QStringLiteral("update_handle")},
            {QStringLiteral("objectId"), current.value(QStringLiteral("objectId")).toString()},
            {QStringLiteral("handleId"), current.value(QStringLiteral("handleId")).toString()},
            {QStringLiteral("point"), finalPoint}
        };
    }
    if (mode == QStringLiteral("dragging_object")) {
        const QVariantMap startPoint = current.value(QStringLiteral("startPoint")).toMap();
        const double dx = finiteNumber(finalPoint.value(QStringLiteral("x")), 0.0) - finiteNumber(startPoint.value(QStringLiteral("x")), 0.0);
        const double dy = finiteNumber(finalPoint.value(QStringLiteral("y")), 0.0) - finiteNumber(startPoint.value(QStringLiteral("y")), 0.0);
        const QVariantList selectedIds = current.value(QStringLiteral("selectedIds")).toList();
        if (selectedIds.size() > 1) {
            return {{QStringLiteral("kind"), QStringLiteral("move_objects")}, {QStringLiteral("objectIds"), selectedIds}, {QStringLiteral("dx"), dx}, {QStringLiteral("dy"), dy}};
        }
        return {{QStringLiteral("kind"), QStringLiteral("move_object")}, {QStringLiteral("objectId"), current.value(QStringLiteral("objectId")).toString()}, {QStringLiteral("dx"), dx}, {QStringLiteral("dy"), dy}};
    }
    if (mode == QStringLiteral("marquee_select")) {
        return {
            {QStringLiteral("kind"), QStringLiteral("select_objects")},
            {QStringLiteral("objectIds"), asStringListVariant(payload.value(QStringLiteral("objectIds")))},
            {QStringLiteral("startPoint"), current.value(QStringLiteral("startPoint")).toMap()},
            {QStringLiteral("endPoint"), finalPoint}
        };
    }
    if (mode == QStringLiteral("panning")) {
        const QVariantMap startScreenPoint = current.value(QStringLiteral("startScreenPoint")).toMap();
        return {
            {QStringLiteral("kind"), QStringLiteral("pan")},
            {QStringLiteral("dxPx"), finiteNumber(finalScreenPoint.value(QStringLiteral("x")), 0.0) - finiteNumber(startScreenPoint.value(QStringLiteral("x")), 0.0)},
            {QStringLiteral("dyPx"), finiteNumber(finalScreenPoint.value(QStringLiteral("y")), 0.0) - finiteNumber(startScreenPoint.value(QStringLiteral("y")), 0.0)}
        };
    }
    if (mode == QStringLiteral("drawing_pending_shape")) {
        return {{QStringLiteral("kind"), QStringLiteral("draw_click")}, {QStringLiteral("point"), finalPoint}};
    }
    return noneIntent();
}

QString finishKind(const QVariantMap &state) {
    const QVariantMap current = cloneState(state);
    if (isHandleDrag(current) || isObjectDrag(current)) {
        return QStringLiteral("incremental_drag");
    }
    if (isMarquee(current)) {
        return current.value(QStringLiteral("moved")).toBool() ? QStringLiteral("marquee_select") : QStringLiteral("marquee_click");
    }
    const QString mode = current.value(QStringLiteral("mode")).toString();
    if (mode == QStringLiteral("panning")) {
        return QStringLiteral("pan");
    }
    if (mode == QStringLiteral("drawing_pending_shape")) {
        return QStringLiteral("draw_click");
    }
    return QStringLiteral("none");
}

} // namespace

QVariantMap initialGestureState() {
    return {
        {QStringLiteral("mode"), QStringLiteral("idle")},
        {QStringLiteral("started"), false},
        {QStringLiteral("objectId"), QString()},
        {QStringLiteral("handleId"), QString()},
        {QStringLiteral("selectedIds"), QVariantList()},
        {QStringLiteral("startPoint"), pointToVariant({0.0, 0.0})},
        {QStringLiteral("lastPoint"), pointToVariant({0.0, 0.0})},
        {QStringLiteral("startScreenPoint"), pointToVariant({0.0, 0.0})},
        {QStringLiteral("lastScreenPoint"), pointToVariant({0.0, 0.0})},
        {QStringLiteral("moved"), false},
        {QStringLiteral("modifiers"), normalizeModifiers({})},
        {QStringLiteral("targetKind"), QStringLiteral("none")},
        {QStringLiteral("targetObjectId"), QString()},
        {QStringLiteral("targetHandleId"), QString()},
        {QStringLiteral("rejected"), false}
    };
}

QVariantMap beginHover(const QVariantMap &state, const CanvasPoint &point, const QVariantMap &target) {
    return beginGesture(state, QStringLiteral("hovering"), {
        {QStringLiteral("point"), pointToVariant(point)},
        {QStringLiteral("targetKind"), target.value(QStringLiteral("kind"), QStringLiteral("none")).toString()},
        {QStringLiteral("targetObjectId"), target.value(QStringLiteral("objectId")).toString()},
        {QStringLiteral("targetHandleId"), target.value(QStringLiteral("handleId")).toString()},
        {QStringLiteral("modifiers"), target.value(QStringLiteral("modifiers")).toMap()}
    });
}

QVariantMap beginObjectDrag(const QVariantMap &state, const QString &objectId, const CanvasPoint &point, const QVariantList &selectedIds, const QVariantMap &modifiers) {
    return beginGesture(state, QStringLiteral("dragging_object"), {{QStringLiteral("objectId"), objectId}, {QStringLiteral("point"), pointToVariant(point)}, {QStringLiteral("selectedIds"), selectedIds}, {QStringLiteral("modifiers"), modifiers}});
}

QVariantMap beginHandleDrag(const QVariantMap &state, const QString &objectId, const QString &handleId, const CanvasPoint &point, const QVariantMap &modifiers) {
    return beginGesture(state, QStringLiteral("dragging_handle"), {{QStringLiteral("objectId"), objectId}, {QStringLiteral("handleId"), handleId}, {QStringLiteral("point"), pointToVariant(point)}, {QStringLiteral("modifiers"), modifiers}});
}

QVariantMap beginMarquee(const QVariantMap &state, const CanvasPoint &point, const QVariantMap &modifiers) {
    return beginGesture(state, QStringLiteral("marquee_select"), {{QStringLiteral("point"), pointToVariant(point)}, {QStringLiteral("modifiers"), modifiers}});
}

QVariantMap beginPan(const QVariantMap &state, const ScreenPoint &screenPoint, const QVariantMap &modifiers) {
    return beginGesture(state, QStringLiteral("panning"), {{QStringLiteral("screenPoint"), QVariantMap{{QStringLiteral("x"), screenPoint.x}, {QStringLiteral("y"), screenPoint.y}}}, {QStringLiteral("modifiers"), modifiers}});
}

QVariantMap beginDrawingPendingShape(const QVariantMap &state, const CanvasPoint &point, const QVariantMap &modifiers) {
    return beginGesture(state, QStringLiteral("drawing_pending_shape"), {{QStringLiteral("point"), pointToVariant(point)}, {QStringLiteral("modifiers"), modifiers}});
}

QVariantMap updateGesture(const QVariantMap &state, const QVariantMap &payload) {
    QVariantMap next = cloneState(state);
    next.insert(QStringLiteral("rejected"), false);
    next.insert(QStringLiteral("modifiers"), normalizeModifiers(payload.contains(QStringLiteral("modifiers")) ? payload.value(QStringLiteral("modifiers")) : next.value(QStringLiteral("modifiers"))));
    if (payload.contains(QStringLiteral("point"))) {
        next.insert(QStringLiteral("lastPoint"), normalizePointMap(payload.value(QStringLiteral("point"))));
    }
    if (payload.contains(QStringLiteral("screenPoint"))) {
        next.insert(QStringLiteral("lastScreenPoint"), normalizeScreenPointMap(payload.value(QStringLiteral("screenPoint"))));
    }
    if (payload.contains(QStringLiteral("targetKind"))) {
        next.insert(QStringLiteral("targetKind"), payload.value(QStringLiteral("targetKind"), QStringLiteral("none")).toString());
        next.insert(QStringLiteral("targetObjectId"), payload.value(QStringLiteral("targetObjectId")).toString());
        next.insert(QStringLiteral("targetHandleId"), payload.value(QStringLiteral("targetHandleId")).toString());
    }
    const QVariantMap lastPoint = next.value(QStringLiteral("lastPoint")).toMap();
    const QVariantMap startPoint = next.value(QStringLiteral("startPoint")).toMap();
    const QVariantMap lastScreenPoint = next.value(QStringLiteral("lastScreenPoint")).toMap();
    const QVariantMap startScreenPoint = next.value(QStringLiteral("startScreenPoint")).toMap();
    const double dx = finiteNumber(lastPoint.value(QStringLiteral("x")), 0.0) - finiteNumber(startPoint.value(QStringLiteral("x")), 0.0);
    const double dy = finiteNumber(lastPoint.value(QStringLiteral("y")), 0.0) - finiteNumber(startPoint.value(QStringLiteral("y")), 0.0);
    const double sdx = finiteNumber(lastScreenPoint.value(QStringLiteral("x")), 0.0) - finiteNumber(startScreenPoint.value(QStringLiteral("x")), 0.0);
    const double sdy = finiteNumber(lastScreenPoint.value(QStringLiteral("y")), 0.0) - finiteNumber(startScreenPoint.value(QStringLiteral("y")), 0.0);
    const double tolerance = std::max(0.0, finiteNumber(payload.value(QStringLiteral("moveTolerance")), 0.000001));
    const double screenTolerance = std::max(0.0, finiteNumber(payload.value(QStringLiteral("screenMoveTolerancePx")), 0.0));
    next.insert(QStringLiteral("moved"), next.value(QStringLiteral("moved")).toBool()
        || std::abs(dx) > tolerance
        || std::abs(dy) > tolerance
        || std::abs(sdx) > screenTolerance
        || std::abs(sdy) > screenTolerance);
    return next;
}

QVariantMap finishGesture(const QVariantMap &state, const QVariantMap &payload) {
    return {
        {QStringLiteral("state"), initialGestureState()},
        {QStringLiteral("intent"), finishIntent(state, payload)}
    };
}

QVariantMap cancelGesture(const QVariantMap &) {
    return {
        {QStringLiteral("state"), initialGestureState()},
        {QStringLiteral("intent"), noneIntent()}
    };
}

QVariantMap finishAction(const QVariantMap &state) {
    const QString kind = finishKind(state);
    return {
        {QStringLiteral("kind"), kind},
        {QStringLiteral("shouldFinishIncrementalDrag"), kind == QStringLiteral("incremental_drag")},
        {QStringLiteral("shouldSelectMarquee"), kind == QStringLiteral("marquee_select")},
        {QStringLiteral("shouldSuppressClick"), kind == QStringLiteral("incremental_drag") || kind == QStringLiteral("marquee_select")},
        {QStringLiteral("shouldEndObjectMove"), kind != QStringLiteral("marquee_select")},
        {QStringLiteral("shouldResetLifecycle"), true}
    };
}

bool isDragging(const QVariantMap &state) {
    const QString mode = state.value(QStringLiteral("mode")).toString();
    return mode == QStringLiteral("dragging_object") || mode == QStringLiteral("dragging_handle");
}

bool isHandleDrag(const QVariantMap &state) {
    return state.value(QStringLiteral("mode")).toString() == QStringLiteral("dragging_handle");
}

bool isObjectDrag(const QVariantMap &state) {
    return state.value(QStringLiteral("mode")).toString() == QStringLiteral("dragging_object");
}

bool isMarquee(const QVariantMap &state) {
    return state.value(QStringLiteral("mode")).toString() == QStringLiteral("marquee_select");
}

QString gestureLabel(const QVariantMap &state) {
    const QString mode = state.value(QStringLiteral("mode"), QStringLiteral("idle")).toString();
    if (mode == QStringLiteral("dragging_handle")) {
        return state.value(QStringLiteral("handleId")).toString() == QStringLiteral("rect_rotate") ? QStringLiteral("rotate") : QStringLiteral("drag handle");
    }
    if (mode == QStringLiteral("dragging_object")) {
        return state.value(QStringLiteral("selectedIds")).toList().size() > 1 ? QStringLiteral("move selection") : QStringLiteral("move object");
    }
    if (mode == QStringLiteral("marquee_select")) {
        return QStringLiteral("marquee");
    }
    if (mode == QStringLiteral("panning")) {
        return QStringLiteral("pan");
    }
    if (mode == QStringLiteral("drawing_pending_shape")) {
        return QStringLiteral("draw");
    }
    return mode == QStringLiteral("hovering") ? QStringLiteral("hover") : QStringLiteral("idle");
}

} // namespace drawing_canvas
